mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,346 @@
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+import loopy as lp
+from modepy import vandermonde
+from pytools.obj_array import obj_array_vectorize
+from meshmode.dof_array import thaw, DOFArray
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+from grudge.eager import interior_trace_pair, cross_rank_trace_pairs
+from grudge.symbolic.primitives import TracePair
+from mirgecom.fluid import (
+    split_conserved,
+    join_conserved_vectors
+)
+
+
+def _facial_flux_q(discr, q_tpair):
+    """Compute facial flux for each scalar component of Q."""
+    q_int = q_tpair.int
+    actx = q_int[0].array_context
+
+    flux_dis = q_tpair.avg
+    if isinstance(flux_dis, np.ndarray):
+        flux_dis = flux_dis.reshape(-1, 1)
+
+    normal = thaw(actx, discr.normal(q_tpair.dd))
+
+    # This uses a central scalar flux along nhat:
+    # flux = 1/2 * (Q- + Q+) * nhat
+    flux_out = flux_dis * normal
+
+    return discr.project(q_tpair.dd, "all_faces", flux_out)
+
+
+def _facial_flux_r(discr, r_tpair):
+    """Compute facial flux for vector compontnent of grad(Q)."""
+    r_int = r_tpair.int
+    actx = r_int[0][0].array_context
+
+    normal = thaw(actx, discr.normal(r_tpair.dd))
+
+    # This uses a central vector flux along nhat:
+    # flux = 1/2 * (grad(Q)- + grad(Q)+) .dot. nhat
+    flux_out = r_tpair.avg @ normal
+
+    return discr.project(r_tpair.dd, "all_faces", flux_out)
+
+
+def av_operator(discr, t, eos, boundaries, q, alpha, **kwargs):
+    r"""Compute artifical viscosity for the euler equations.
+
+    Computes the the right-hand-side term for artificial viscosity.
+
+    .. math::
+
+        \mbox{RHS}_{\mbox{av}} = \nabla\cdot{\varepsilon\nabla\mathbf{Q}}
+
+    Parameters
+    ----------
+    q
+        State array which expects the quantity to be limited on to be listed
+        first in the array. For the Euler equations this could be the canonical
+        conserved variables (mass, energy, mometum) for the fluid along with a
+        vector of species masses for multi-component fluids.
+
+    boundaries
+        Dicitionary of boundary functions, one for each valid boundary tag
+
+    t
+        Time
+
+    alpha
+       The maximum artifical viscosity coeffiecent to be applied
+
+    eos: mirgecom.eos.GasEOS
+       Only used as a pass through to the boundary conditions.
+
+    Returns
+    -------
+    numpy.ndarray
+        Agglomerated object array of DOF Arrays representing the RHS associated
+        with the artificial viscosity application.
+    """
+    dim = discr.dim
+    cv = split_conserved(dim, q)
+
+    # Get smoothness indicator based on fluid mass density
+    indicator = smoothness_indicator(discr, cv.mass, **kwargs)
+
+    # R=Grad(Q) volume part
+    grad_q_vol = join_conserved_vectors(dim, obj_array_vectorize(discr.weak_grad, q))
```

Comment:
No.  It does not work as you want it to when doing grad(V) where V is a vector.

Reply:
Oh wait... yeah you might be right. I think it might already do the right thing here without `obj_array_vectorize`. Q is not a vector quantity like those that fail when used with `weak_grad`.

Reply:
inducer/grudge#87